pkg/services/notary/notary.go

```diff
@@ -0,0 +1,364 @@
+			Chain:   bc,
+			Log:     log,
+		},
+		wallet: wallet,
+	}, nil
+}
+
+// OnNewRequest is a callback method which is called after new notary request is added to the notary request pool.
+func (n *Notary) OnNewRequest(payload *payload.P2PNotaryRequest) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	var isValid bool
+	nvbFallback := payload.FallbackTransaction.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height
+	nKeys := payload.MainTransaction.GetAttributes(transaction.NotaryAssistedT)[0].Value.(*transaction.NotaryAssisted).NKeys
+	typ, nSigs, pubs, err := n.verifyIncompleteWitnesses(payload.MainTransaction, nKeys)
+	if err == nil { // valid main tx signatures
+		isValid = true
+	}
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	r, exists := n.requests[payload.MainTransaction.Hash()]
+	if exists {
+		if r.nSigsCollected >= nSigs { // already collected sufficient number of signatures to complete main transaction
+			return // TODO: however, we may still accept extra fallback transactions without adding new signatures to main tx
+		}
+		r.isValid = isValid && r.isValid
+		if nvbFallback < r.minNotValidBefore {
+			r.minNotValidBefore = nvbFallback
+		}
+	} else {
+		r = request{
+			nSigs:             nSigs,
+			main:              payload.MainTransaction,
+			typ:               typ,
+			isValid:           isValid,
+			minNotValidBefore: nvbFallback,
+		}
+	}
+	if r.isValid {
+	loop:
+		for i, w := range payload.MainTransaction.Scripts {
+			if len(w.InvocationScript) != 0 {
+				switch r.typ {
+				case Signature:
+					if !exists || len(r.main.Scripts[i].InvocationScript) == 0 { // need this check because signature can already be added (consider receiving the same payload multiple times)
+						r.main.Scripts[i] = w
+						r.nSigsCollected++
+					}
+					if r.nSigsCollected == r.nSigs {
+						break loop
+					}
+				case MultiSignature:
+					if r.sigs == nil {
+						r.sigs = make(map[*keys.PublicKey][]byte)
+					}
+
+					hash := r.main.GetSignedHash().BytesBE()
+					for _, pub := range pubs {
+						if _, ok := r.sigs[pub]; ok {
+							continue // signature for this pub has already been added
+						}
+						if pub.Verify(w.InvocationScript[2:], hash) { // then pub is the owner of the signature
+							r.sigs[pub] = w.InvocationScript
+							r.nSigsCollected++
+							if r.nSigsCollected == r.nSigs {
+								var invScript []byte
+								for j := range pubs {
+									if sig, ok := r.sigs[pubs[j]]; ok {
+										invScript = append(invScript, sig...)
+									}
+								}
+								r.main.Scripts[i].InvocationScript = invScript
+							}
+							break loop
+						}
+					}
+					// pubKey was not found for the signature i.e. signature is bad - we're OK with that, let the fallback TX to be added
+					break loop // only one multisignature is allowed
+				}
+			}
+		}
+	}
+	var isPresent bool
+	for _, f := range r.fallbacks {
+		if f.Hash().Equals(payload.FallbackTransaction.Hash()) {
```

Comment:
BTW, if you already have this fallback you've already processed this payload, so I think this check can be done even before `loop`.

Reply:
Fixed, it makes things a little simplier.